Zend/zend_hash.c

```diff
@@ -244,6 +244,26 @@ ZEND_API HashTable* ZEND_FASTCALL _zend_new_array(uint32_t nSize)
 	return ht;
 }
 
+ZEND_API HashTable* ZEND_FASTCALL zend_new_pair(zval *val1, zval *val2)
+{
+	Bucket *p;
+	HashTable *ht = emalloc(sizeof(HashTable));
```

Comment:
How aggressive do you want to be here ? 

Might it be possible to make alloc for ht+buckets one allocation, might it also be possible to omit hash_init_int, and write a hash_init_pair ?

Reply:
Would love to combine those allocations, but I don't think it's possible. The rest of the hash API assumes that these are separate allocations, and in the end the pair is still a normal array and can be used as such.

Reply:
That's a shame ... what about a specialized init ? among other things, init_int sets flags and then immediately real_init_packed sets them otherwise ? it looks like you could avoid the persistent branch in real_init too ?

Reply:
I would hope that all of this will get optimized because these functions are always_inline, but I'll have to check assembly to be sure.